### DISCUSSION TITLE
gcloud Application Default Credentials are not mounted from `/.config`

### DISCUSSION THREAD

**Discussion by shahar1** [apache/airflow#35469] [Q&A]
I'm trying to get back to Airflow development after a long while, running on WSL2 (Debian 11) with breeze setup.
After pulling the latest changes and rebuilding the Docker images using breeze, I ran an artbitrary GCP operator and got the following credentials error:
```text
[2023-11-06, 06:16:23 UTC] {taskinstance.py:2619} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/opt/airflow/airflow/models/taskinstance.py", line 432, in _execute_task
    result = execute_callable(context=context, **execute_callable_kwargs)
  File "/opt/airflow/airflow/providers/google/cloud/operators/gcs.py", line 260, in execute
    project_id=hook.project_id,
  File "/opt/airflow/airflow/providers/google/common/hooks/base_google.py", line 361, in project_id
    _, project_id = self.get_credentials_and_project_id()
  File "/opt/airflow/airflow/providers/google/common/hooks/base_google.py", line 273, in get_credentials_and_project_id
    credentials, project_id = get_credentials_and_project_id(
  File "/opt/airflow/airflow/providers/google/cloud/utils/credentials_provider.py", line 362, in get_credentials_and_project_id
    return _CredentialProvider(*args, **kwargs).get_credentials_and_project()
  File "/opt/airflow/airflow/providers/google/cloud/utils/credentials_provider.py", line 243, in get_credentials_and_project
    credentials, project_id = self._get_credentials_using_adc()
  File "/opt/airflow/airflow/providers/google/cloud/utils/credentials_provider.py", line 348, in _get_credentials_using_adc
    credentials, project_id = google.auth.default(scopes=self.scopes)
  File "/usr/local/lib/python3.8/site-packages/google/auth/_default.py", line 691, in default
    raise exceptions.DefaultCredentialsError(_CLOUD_SDK_MISSING_CREDENTIALS)
google.auth.exceptions.DefaultCredentialsError: Your default credentials were not found. To set up Application Default Credentials, see https://cloud.google.com/docs/authentication/external/set-up-adc for more information.
```

I checked, and the connection URI seems to be defined properly in `files/airflow-breeze-config/variables.env` (it has always been like that and worked ok until now):
```
AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT='google-cloud-platform://'
```

However - I've figured that the gcloud Application Default Credentials (ADC) json file, as well as the `gcloud` folder, are not mounted from `~/.config`.
When running `ls -ln` on the host's `~/.config` folder I get the following information:
`drwxr-xr-x 5 1000 1000 4096 Nov  6 09:04 gcloud`
I tried to run `breeze ci fix-ownership`, but it didn't help.
When I created the file and copied the contents manually within the Docker container - it worked. Defining the connection directly in Airflow's UI also worked. In any case, it isn't very pleasant to work in either ways :)

I'd be happy for suggestions. Thanks!

**Comment by shahar1** [comment]
That was a silly one - I forgot to pass `--forward-credentials` 🤦‍♂️

**Reply by kunaljubce** [reply]
@shahar1 Can you please help with the layout of the `files/airflow-breeze-config/variables.env` file? I have set it up with the below code from https://github.com/apache/airflow/blob/main/contributing-docs/testing/system_tests.rst

```python
# Build variables. This file is sourced by Breeze.
# Also it is sourced during continuous integration build in Cloud Build

# Auto-export all variables
set -a

echo
echo "Reading variables"
echo

# Generate random number that will be used across your session
RANDOM_FILE="/random.txt"

if [[ ! -f "${RANDOM_FILE}" ]]; then
    echo "${RANDOM}" > "${RANDOM_FILE}"
fi

RANDOM_POSTFIX=$(cat "${RANDOM_FILE}")
```

However, once I start breeze, I still have to manually export variables like `GOOGLE_CLOUD_PROJECT`, `SYSTEM_TESTS_GCP_PROJECT`, `SYSTEM_TESTS_ENV_ID`.

**Reply by potiuk** [reply]
The `variables.env` should be just `VAR=VALUE`. What you are looking at is `init.sh` file in the same directory. See https://github.com/apache/airflow/blob/main/dev/breeze/doc/02_customizing.rst#customizing-breeze-environment for details

**Reply by shahar1** [reply]
Specifically I have the following defined as env. variables in`variables.env`, so it will recognize my [application default credentials](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/connections/gcp.html#note-on-application-default-credentials):
```bash
AIRFLOW_CONN_GOOGLE_CLOUD_DEFAULT='google-cloud-platform://'
GOOGLE_CLOUD_PROJECT='my-gcp-project-id'
```
